pkg/streamingpromql/operators/aggregations/topkbottomk/instant_query.go

```diff
@@ -0,0 +1,320 @@
+
+	groupLabelsBytesFunc := t.groupLabelsBytesFunc()
+	groups := map[string]*instantQueryGroup{}
+	seriesToGroups := make([]*instantQueryGroup, 0, len(innerSeries))
+
+	for _, series := range innerSeries {
+		groupLabelsString := groupLabelsBytesFunc(series.Labels)
+		g, groupExists := groups[string(groupLabelsString)] // Important: don't extract the string(...) call here - passing it directly allows us to avoid allocating it.
+
+		if !groupExists {
+			g = &instantQueryGroup{}
+			groups[string(groupLabelsString)] = g
+		}
+
+		g.seriesCount++
+		seriesToGroups = append(seriesToGroups, g)
+	}
+
+	outputSeriesCount := 0
+
+	for idx, series := range innerSeries {
+		g := seriesToGroups[idx]
+
+		data, err := t.Inner.NextSeries(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		if len(data.Histograms) > 0 {
+			t.emitIgnoredHistogramsAnnotation()
+		}
+
+		if len(data.Floats) == 0 {
+			// No float values, nothing to do.
+			g.seriesCount-- // Avoid allocating space for this series if we can.
+			types.PutInstantVectorSeriesData(data, t.MemoryConsumptionTracker)
+			continue
+		}
+
+		if t.accumulateValue(series, data.Floats[0].F, g) {
+			outputSeriesCount++
+		}
+
+		types.PutInstantVectorSeriesData(data, t.MemoryConsumptionTracker)
+	}
+
+	outputSeries := types.GetSeriesMetadataSlice(outputSeriesCount)
+	t.values, err = types.Float64SlicePool.Get(outputSeriesCount, t.MemoryConsumptionTracker)
+	if err != nil {
+		return nil, err
+	}
+
+	outputSeries = outputSeries[:outputSeriesCount]
+	t.values = t.values[:outputSeriesCount]
+	firstOutputSeriesIndexForNextGroup := 0
+
+	for _, g := range groups {
+		t.heap.Reset(g)
+
+		lastOutputSeriesIndexForGroup := firstOutputSeriesIndexForNextGroup + len(g.series) - 1
+		nextOutputSeriesIndex := lastOutputSeriesIndexForGroup - g.nanCount
+		firstOutputSeriesIndexForNextGroup += len(g.series)
+
+		// Pop returns the next lowest value (topk) or next highest value (bottomk), but we want to return values
+		// in descending order for topk / ascending order for bottomk. The only exception is NaN values, which
+		// should always be last, regardless of the sort order.
+		for len(g.series) > 0 {
+			next := heap.Pop(t.heap).(instantQuerySeries)
+
+			if math.IsNaN(next.value) {
+				idx := lastOutputSeriesIndexForGroup - g.nanCount + 1
+				outputSeries[idx] = next.metadata
+				t.values[idx] = next.value
+				g.nanCount--
+			} else {
+				outputSeries[nextOutputSeriesIndex] = next.metadata
+				t.values[nextOutputSeriesIndex] = next.value
+				nextOutputSeriesIndex--
+			}
+		}
+	}
+
+	return outputSeries, nil
+}
+
+func (t *InstantQuery) loadLimit(ctx context.Context) error {
+	paramValues, err := t.Param.GetValues(ctx)
+	if err != nil {
+		return err
+	}
+
+	defer types.FPointSlicePool.Put(paramValues.Samples, t.MemoryConsumptionTracker)
+
+	v := paramValues.Samples[0].F
+
+	if !convertibleToInt64(v) {
+		return fmt.Errorf("scalar value %v overflows int64", v)
```

Comment:
```suggestion
		return fmt.Errorf("scalar value %v overflows int64 for use in `top/bottomk`", v)
```

What does Prometheus do here?

Reply:
Okay. Can this message be improved in MQE? We shouldn't need to be identical here right?

Reply:
Yep, fair enough, changed in 1c856c4